src/ts/spatialnavigation/spatialnavigation.ts

```diff
@@ -0,0 +1,170 @@
+import { NavigationGroup } from './navigationgroup';
+import { RootNavigationGroup } from './rootnavigationgroup';
+import { NodeEventSubscriber } from './nodeeventsubscriber';
+import { SeekBarHandler } from './seekbarhandler';
+
+export enum Direction {
+  UP = 'up',
+  DOWN = 'down',
+  LEFT = 'left',
+  RIGHT = 'right',
+}
+
+export enum Action {
+  SELECT = 'select',
+  BACK = 'back',
+}
+
+const KeyMap: {[key: number]: Direction | Action} = {
+  4: Direction.LEFT,
+  21: Direction.LEFT,
+  37: Direction.LEFT,
+  214: Direction.LEFT,
+  205: Direction.LEFT,
+  218: Direction.LEFT,
+  5: Direction.RIGHT,
+  22: Direction.RIGHT,
+  39: Direction.RIGHT,
+  213: Direction.RIGHT,
+  206: Direction.RIGHT,
+  217: Direction.RIGHT,
+  29460: Direction.UP,
+  19: Direction.UP,
+  38: Direction.UP,
+  211: Direction.UP,
+  203: Direction.UP,
+  215: Direction.UP,
+  29461: Direction.DOWN,
+  20: Direction.DOWN,
+  40: Direction.DOWN,
+  212: Direction.DOWN,
+  204: Direction.DOWN,
+  216: Direction.DOWN,
+  29443: Action.SELECT,
+  13: Action.SELECT,
+  67: Action.SELECT,
+  32: Action.SELECT,
+  23: Action.SELECT,
+  195: Action.SELECT,
+  27: Action.BACK,
+};
+
+export class Spatialnavigation {
+  private unsubscribeVisibilityChangesFns: (() => void)[];
+  private readonly navigationGroups: NavigationGroup[] = [];
+  private readonly activeNavigationGroups: NavigationGroup[];
+  private readonly eventSubscriber: NodeEventSubscriber;
+  private readonly seekBarHandler: SeekBarHandler;
+
+  constructor(rootNavigationGroup: RootNavigationGroup, ...navigationGroups: NavigationGroup[]) {
+    this.seekBarHandler = new SeekBarHandler(rootNavigationGroup);
+
+    this.activeNavigationGroups = [];
+    this.unsubscribeVisibilityChangesFns = [];
+    this.eventSubscriber = new NodeEventSubscriber();
+    this.navigationGroups = [rootNavigationGroup, ...navigationGroups];
+
+    this.subscribeToNavigationGroupVisibilityChanges();
+    this.attachKeyEventHandler();
+    this.enableDefaultNavigationGroup();
+  }
+
+  private attachKeyEventHandler(): void {
+    this.eventSubscriber.on(document, 'keydown', this.handleKeyEvent, true);
+  }
+
+  private onShow = (group: NavigationGroup): void => {
+    this.activeNavigationGroups.push(group);
+    this.updateEnabledNavigationGroup();
+  };
+
+  private onHide = (group: NavigationGroup): void => {
+    const groupIndex = this.activeNavigationGroups.findIndex(other => other === group);
+
+    if (groupIndex > -1) {
+      group.disable();
+      this.activeNavigationGroups.splice(groupIndex, 1);
+      this.updateEnabledNavigationGroup();
+    }
+  };
+
+  private subscribeToNavigationGroupVisibilityChanges(): void {
+    this.navigationGroups.forEach(group => {
+      const onShowHandler = this.onShow.bind(this, group);
+      const onHideHandler = this.onHide.bind(this, group);
```

Comment:
Do we _need_ to bind these? They are declared as arrow functions, that should retain the scope right?

Reply:
No, we don't 😄 Changed in da004f28 - using error functions instead now.